pptx/opc/package.py

```diff
@@ -104,17 +108,15 @@ def main_document_part(self):
 
     def next_partname(self, tmpl):
         """
-        Return a |PackURI| instance representing the next available partname
+        Return a |PackURI| instance representing the next partname
         matching *tmpl*, which is a printf (%)-style template string
         containing a single replacement item, a '%d' to be used to insert the
         integer portion of the partname. Example: '/ppt/slides/slide%d.xml'
         """
-        partnames = [part.partname for part in self.iter_parts()]
-        for n in range(1, len(partnames) + 2):
-            candidate_partname = tmpl % n
-            if candidate_partname not in partnames:
-                return PackURI(candidate_partname)
-        raise Exception("ProgrammingError: ran out of candidate_partnames")
+        name = tmpl.split(os.sep)[2]
+        self.partnames[name] += 1
+        candidate_partname = tmpl % self.partnames[name]
+        return PackURI(candidate_partname)
 
```

Comment:
Where are the tests? You would have found most of these problems with good tests.